### PR TITLE
Removed NoSourceCode at GetDeclarationLocation 

### PR DESCRIPTION
**Pull request by amieres**
Reintroducing PR rebased to latest master
https://github.com/Microsoft/visualfsharp/pull/3719

Removed check for NoSourceCode error from TypeCheckInfo.GetDeclarationLocation.
It was checking for a file to exist and failing if it did not.

fsharp/FSharp.Compiler.Service#821

### CODE REVIEW

**Comment by cartermp** [comment]
It's that damn "Test Project21 whole project errors" error that shows up in every PR again...

```
16:33:24 => Tests.Service.ProjectAnalysisTests.Test Project21 whole project errors
16:33:24 Project21 error: <<<This expression was expected to have type
16:33:24     'int'    
16:33:24 but here has type
16:33:24     'string'    >>>
16:33:24 Project21 error: <<<Invalid value>>>
```

@dsyme any idea what this test is for?

**Comment by amieres** [comment]
There are also some 6 tests that are failing, all related to a TypeProvider:

These are the failing tests:
```
GotoDefinition.ProvidedTypeNoDefinitionLocationAttribute
GotoDefinition.TypeProvider.DefinitionLocationAttribute.Constructor.FileDoesnotExist
GotoDefinition.TypeProvider.DefinitionLocationAttribute.Event.FileDoesnotExist
GotoDefinition.TypeProvider.DefinitionLocationAttribute.Method.FileDoesnotExist
GotoDefinition.TypeProvider.DefinitionLocationAttribute.Property.FileDoesnotExist
GotoDefinition.TypeProvider.DefinitionLocationAttribute.Type.FileDoesnotExist
```

and messages look like this: 
```
Expected GotoDefn to fail, but it went to a definition in file C:\Users\dotnet-bot\AppData\Local\Temp\salsa-1911234-4984\Temp.fs]]

at Salsa.VsOpsUtils.CheckGotoDefnResult(FSharpOption`1 expected, OpenFile file, GotoDefinitionResult_DEPRECATED actual) in D:\j\w\release_ci_pa---3f142ccc\vsintegration\tests\Salsa\SalsaUtils.fs:line 332
at <StartupCode$VisualFSharp-UnitTests>.$Tests.LanguageService.GotoDefinition.VerifyGoToDefnFailAtStartOfMarker@50.Invoke(Tuple`2 tupledArg) in D:\j\w\release_ci_pa---3f142ccc\vsintegration\tests\unittests\LegacyLanguageService\Tests.LanguageService.GotoDefinition.fs:line 0
at Tests.LanguageService.GotoDefinition.UsingMSBuild.VerifyGoToDefnFailAtStartOfMarker(String fileContents, String marker, FSharpFunc`2 f, FSharpOption`1 addtlRefAssy) in D:\j\w\release_ci_pa---3f142ccc\vsintegration\tests\unittests\LegacyLanguageService\Tests.LanguageService.GotoDefinition.fs:line 61
```

**Comment by amieres** [comment]
I think I found what the issue was. The clue was in this comment:
```
// provided items may have TypeProviderDefinitionLocationAttribute that binds them to some location
```
I realized that the file check was probably intended precisely to reject those TypeProvider generated items. 
I changed a larger section and now I am simply failing GetDeclarationLocation whenever the item is TypeProvider generated, without ever checking for a file.

**Comment by amieres** [comment]
Finally all tests passed!
Ready to merge!

File: src/fsharp/service/service.fs [hunk 1]
```diff
@@ -1214,32 +1214,28 @@ type TypeCheckInfo
...
               | Some x -> x
-              | None ->
-                  let fail defaultReason = 
-                      match item.Item with 
+              | None   ->
+              match rangeOfItem g preferFlag item.Item with
+              | Some itemRange -> 
+                  let projectDir = Filename.directoryName (if projectFileName = "" then mainInputFileName else projectFileName)
+                  fileNameOfItem g (Some projectDir) itemRange item.Item
+                  |> mkRange <| itemRange.Start <| itemRange.End
```
Comment (dsyme) [inline src/fsharp/service/service.fs]: Please never use the ``<|`` operator in the compiler code any more than it is today

**Comment by amieres** [comment]
Removed the `<|`

**Comment by cartermp** [comment]
Thanks @amieres!

**Comment by amieres** [comment]
my pleasure!

**Pull request merged**